### PR TITLE
Add function to check whether a point is in the user-defined domain

### PR DESCRIPTION
**Pull request by anne-glerum**
As suggested in #1044, a function that checks whether a given point falls within the boundaries specified in the input file. @jdannberg is this what you had in mind?

### CODE REVIEW

File: source/geometry_model/box.cc [hunk 1]
```diff
@@ -189,6 +190,32 @@ namespace aspect
     }
 
     template <int dim>
+    bool
+    Box<dim>::point_is_in_domain(const Point<dim> &point) const
+    {
+      AssertThrow(this->get_free_surface_boundary_indicators().size() == 0 ||
+                  this->get_timestep_number() == 0,
+                  ExcMessage("After displacement of the free surface, this function cannot be used to determine whether point lies in domain or not."));
```
Comment (bangerth) [inline source/geometry_model/box.cc]: -> in THE domain

File: source/geometry_model/box.cc [hunk 2]
```diff
@@ -189,6 +190,32 @@ namespace aspect
...
+
+      AssertThrow(dynamic_cast<const InitialTopographyModel::ZeroTopography<dim>*>(&this->get_initial_topography_model()) != 0,
+                  ExcMessage("After adding topography, this function can no longer be used to determine whether point lies in domain or not."));
```
Comment (bangerth) [inline source/geometry_model/box.cc]: -> in THE domain
same elsewhere as appropriate

File: source/geometry_model/box.cc [hunk 3]
```diff
@@ -189,6 +190,32 @@ namespace aspect
...
+
+      bool in_domain = true;
+
+      for (unsigned int d = 0; d < dim; d++)
+        {
+          if (point[d] > extents[d]+box_origin[d]+std::numeric_limits<double>::epsilon()*extents[d] ||
+              point[d] < box_origin[d]-std::numeric_limits<double>::epsilon()*extents[d])
+            {
+              in_domain = false;
+              break;
```
Comment (bangerth) [inline source/geometry_model/box.cc]: You don't need the `in_domain` variable if you just say `return false` here and then say `return true;` at the bottom of the function (where you only get if every check has succeeded). Same for all of the other similar functions.

File: source/geometry_model/ellipsoidal_chunk.cc [hunk 4]
```diff
@@ -193,6 +195,7 @@ namespace aspect
       coarse_grid.signals.post_refinement.connect(std_cxx11::bind (&EllipsoidalChunk<dim>::set_boundary_ids,
                                                                    std_cxx11::cref(*this),
                                                                    std_cxx11::ref(coarse_grid)));
+
```
Comment (bangerth) [inline source/geometry_model/ellipsoidal_chunk.cc]: no need for the empty line :-)

File: source/geometry_model/interface.cc [hunk 5]
```diff
@@ -65,6 +66,15 @@ namespace aspect
 
 
     template <int dim>
+    bool
+    Interface<dim>::point_is_in_domain(const Point<dim> &p) const
+    {
+      // return false in the base class
+      return false;
```
Comment (bangerth) [inline source/geometry_model/interface.cc]: That seems like the wrong approach. I would either throw an exception here, or just make the function abstract in the base class.

**Comment by bangerth** [comment]
Nice work. Out of curiosity, where do you need this?

**Comment by anne-glerum** [comment]
In #1044 I use a user-defined point to determine where to calculate the pressure profile to prescribe as traction boundary condition. I first check that this point lies in the domain. Juliane suggested instead of doing this check in the traction plugin, to just add a function to the geometry models, as the different logic for all the geometries was cluttering the traction plugin.

**Comment by bangerth** [comment]
Yes, this is definitely simpler.

**Comment by bangerth** [comment]
I was about to merge this, but then remembered that I still had one question. Semantically, I believe what you want this function to do is to test whether the point is inside the domain *as specified by the geometry model by itself*, i.e., without taking into account any possibly applied surface topography (initial or dynamic). Is this correct? If so, would you be willing to add such a sentence (that it does not take into account initial or dynamic surface topographies) to the description of the member functions?

Or do you *intend* for these functions to take this into account, but it hasn't been implemented yet?

**Comment by anne-glerum** [comment]
Indeed I want to test only whether the point is inside the domain as specified by the geometry. I've adapted the member function description.

**Pull request merged**